src/merkle_tree/constraints.rs

```diff
@@ -101,39 +45,166 @@ where
+    ConstraintF: Field,
+{
+    /// Given a leaf, calculate the root as if the merkle has this leaf at this path.
+    fn calculate_root(
+        &self,
+        leaf_hash_params: &LeafH::ParametersVar,
+        two_to_one_hash_params: &TwoToOneH::ParametersVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<TwoToOneH::OutputVar, SynthesisError> {
+        let leaf_bytes = leaf.to_bytes()?;
+        let claimed_leaf_hash = LeafH::evaluate(leaf_hash_params, &leaf_bytes)?;
+        let leaf_sibling_hash = &self.leaf_sibling;
+
+        // calculate hash for the bottom non_leaf_layer
+
+        // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+        // and the bit being 1 indicates our currently hashed value is on the right.
+        // Thus `left_hash` is sibling if leaf_position_bit is 1, and leaf if bit is 0
+
+        let left_hash = self
+            .leaf_is_right_child
+            .select(leaf_sibling_hash, &claimed_leaf_hash)?
+            .to_bytes()?;
+        let right_hash = self
+            .leaf_is_right_child
+            .select(&claimed_leaf_hash, leaf_sibling_hash)?
+            .to_bytes()?;
+
+        let mut curr_hash = TwoToOneH::evaluate(two_to_one_hash_params, &left_hash, &right_hash)?;
+        // To traverse up a MT, we iterate over the path from bottom to top (in reverse)
+
+        // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+        // and the bit being 1 indicates our currently hashed value is on the right.
+        // Thus `left_hash` is the sibling if bit is 1, and it's the computed hash if bit is 0
+        for (bit, sibling) in self.path.iter().rev().zip(self.auth_path.iter().rev()) {
+            let left_hash = bit.select(sibling, &curr_hash)?;
+            let right_hash = bit.select(&curr_hash, sibling)?;
+
+            curr_hash = TwoToOneH::evaluate(
+                two_to_one_hash_params,
+                &left_hash.to_bytes()?,
+                &right_hash.to_bytes()?,
+            )?;
+        }
+
+        Ok(curr_hash)
+    }
+
+    /// Given a leaf, calculate the hash of the merkle tree
+    /// along the path and check if the given root is correct.
+    pub fn verify_membership(
+        &self,
+        leaf_hash_params: &LeafH::ParametersVar,
+        two_to_one_hash_params: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<Boolean<ConstraintF>, SynthesisError> {
+        let expected_root = self.calculate_root(leaf_hash_params, two_to_one_hash_params, leaf)?;
+        // enforce `expected_root` is equal to the given root
+        Ok(expected_root.is_eq(root)?)
+    }
+
+    /// update the `old_leaf` to `new_leaf` and returned the updated MT root. `self` will not be
+    /// modified.
+    ///
+    /// If the `old_leaf` does not lead to `old_root`, constraints will not be satisfied.
```

Comment:
```suggestion
    /// Check that `old_leaf` is the leaf of the Merkle tree on the path defined by 
    /// `self`, and then compute the new root when replacing `old_leaf` by `new_leaf`.
```

I am not sure what this API should do if the `old_leaf` is a dummy one (eg, if we're inserting a new leaf for the first time).

Thoughts @ValarDragon?

Reply:
I added `old_leaf` as a parameter because the code checks if `old_leaf`'s membership before updating the new leaf. 

A better solution might be removing the `old_leaf` parameter and let user decide if it's necessary to check the old_leaf's membership first. Make sense? @Pratyush

Reply:
Hmm not sure

Reply:
cc @ValarDragon

Reply:
Maybe we should have a method for `insert` if the old_leaf is a dummy one (not already included), and otherwise have it called `update`, which works as is?